ibis/file/client.py

```diff
@@ -0,0 +1,196 @@
+                    if str(d).endswith(self.extension):
+                        tables.append(d.stem)
+        return tables
+
+
+class CSVClient(Client):
+    extension = 'csv'
+
+    def table(self, name, path=None):
+        if name not in self.list_tables(path):
+            raise AttributeError(name)
+
+        if path is None:
+            path = self.root
+
+        # get the schema
+        f = path / "{}.{}".format(name, self.extension)
+        df = pd.read_csv(str(f), header=0, nrows=10)
+        df['time'] = pd.to_datetime(df['time']).dt.tz_localize('UTC')
+        schema = pandas_dtypes_to_ibis_schema(df, {})
+
+        t = CSVTable(name, schema, self).to_expr()
+        self.dictionary[name] = f
+        return t
+
+    def list_tables(self, path=None):
+        # tables are files in a dir
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_file():
+                    if str(d).endswith(self.extension):
+                        tables.append(d.stem)
+        return tables
+
+    def list_databases(self, path=None):
+        # databases are dir
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_dir():
+                    tables.append(d.name)
+        return tables
+
+
+class Database(ibis.client.Database):
+
+    def __init__(self, name, client, path=None):
+        super().__init__(name, client)
+        self.path = path
+
+    def __str__(self):
+        return '{0.__class__.__name__}({0.name})'.format(self)
+
+    def __dir__(self):
+        dbs = self.list_databases(path=self.path)
+        tables = self.list_tables(path=self.path)
+        return sorted(list(set(dbs).union(set(tables))))
+
+    def __getattr__(self, name):
+        try:
+            return object.__getattribute__(self, name)
+        except AttributeError:
+            try:
+                return self.table(name, path=self.path)
+            except AttributeError:
+                return self.database(name, path=self.path)
+
+    def table(self, name, path):
+        return self.client.table(name, path=path)
+
+    def database(self, name=None, path=None):
+        return self.client.database(name=name, path=path)
+
+    def list_databases(self, path=None):
+        if path is None:
+            path = self.path
+        return self.client.list_databases(path=path)
+
+    def list_tables(self, path=None):
+        if path is None:
+            path = self.path
+        return self.client.list_tables(path=path)
+
+
+@data_preload.register(HDFTable, pathlib.Path)
+def hdf_data_preload_uri_client(table, path, scope=None, **kwargs):
+    return pd.read_hdf(str(path), table.name, mode='r')
+
+
+@data_preload.register(CSVTable, pathlib.Path)
+def csv_data_preload_uri_client(table, path, scope=None, **kwargs):
```

Comment:
here we could pretty easily push down a projection via ``usecols``

Reply:
note we also assume ``index=False`` and a header in row 0 (as this matches how we write csvs). This could of course be more general, but makes sense to limit it for the easy case (and of course easy enough to simply allow passing in kwargs anyhow)